json: Add runnable for `package.json` and `composer.json` scripts

RemcoSmitsDev opened the pull request:
**Package.json**

https://github.com/zed-industries/zed/assets/62463826/f8ca12a5-1292-4465-83e1-3c2ab65f5833

**Composer.json**

https://github.com/zed-industries/zed/assets/62463826/61f9e74d-c6ed-4329-855b-d0161e0a117b

Release Notes:

- Added runnable for `package.json` and `composer.json` scripts ([#12215](https://github.com/zed-industries/zed/issues/12215)).

RemcoSmitsDev committed:
Add runnable for `package.json` scripts

znycheporuk commented:
To change runner from npm to yarn, for example. Do I have to copy task and replace npm run with yarn, or is it possible, to configure runner somehow?

RemcoSmitsDev committed:
Fix false positives

Before if you would open the task modal you would see a task for `scripts` which is incorrect, because you cannot run it. So changing it to a custom variable fixes this.

RemcoSmitsDev commented:
@znycheporuk @Angelk90 

See the following comment on my JavaScript runnable PR https://github.com/zed-industries/zed/pull/12118#issuecomment-2129232114. We want to make something like that, but I'm not sure if we want to merge this and first just let people duplicate the task if they want to use a different runner or make it more user-friendly first.

RemcoSmitsDev commented:
@osiewicz It might be cool to be able to specify a filename pattern that the task is allowed to show up. This way, we could separate `package.json` and `composer.json` scripts and not show the runnable when your JSON file contains these keys.

RemcoSmitsDev committed:
Remove value query

This allows us to use the same query for `package.json` and `composer.json`

RemcoSmitsDev committed:
Add task for running `composer.json` scriipt

RemcoSmitsDev commented on crates/languages/src/json/runnables.scm:
```diff
@@ -0,0 +1,20 @@
+                )
+            )
+        )
+    ) @package-script
```
Not sure about the name, would love to name them `package-script` and `composer-script`. But not sure what you think about having double capture names for the same capture(only for a different name)

osiewicz replied:
I think this is the best way forward, yeah; However, it doesn't looks like that capture is being used anywhere. We can probably make it anonymous;
```suggestion
    ) @_script
```
Did you mean that it would be cool to have two separate tags for this? If so, that'd be lovely. You can just use `#set! tag` twice with different arguments and it'll just work.

znycheporuk commented:
@RemcoSmitsDev,  I agree that it would be cool to specify env variables for tasks like this
```
tasks:
  javascript:
    TEST_RUNNER: jest
```
In my opinion if adding this functionality will take an additional week or two, it's probably better to wait until it can be implemented in a more general way. However, if there's no clear timeline for when it will be possible to use such variables in tasks, then merge the PR as it is and update the task later.

osiewicz commented:
I'm out in the latter half of the next week, so I'd rather have us start with something (even if it's not ideal) and then refine it once user-provided task variables do land.

osiewicz commented:
> @osiewicz It might be cool to be able to specify a filename pattern that the task is allowed to show up. This way, we could separate package.json and composer.json scripts and not show the runnable when your JSON file contains these keys.

I think that would be feasible to do, yeah

RemcoSmitsDev committed:
Remove unused capture

RemcoSmitsDev committed:
Add `composer-script` script tag

osiewicz reviewed (approved):
LGTM :)

RemcoSmitsDev commented:
Should be good to go now!
